url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/SSLSocketChannel2.java

```diff
@@ -0,0 +1,380 @@
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLEngineResult.Status;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+
+/**
+ * Implements the relevant portions of the SocketChannel interface with the SSLEngine wrapper.
+ */
+public class SSLSocketChannel2 implements ByteChannel, WrappedByteChannel {
+	/**
+	 * This object is used to feed the {@link SSLEngine}'s wrap and unwrap methods during the handshake phase.
+	 **/
+	protected static ByteBuffer emptybuffer = ByteBuffer.allocate( 0 );
+
+	protected ExecutorService exec;
+
+	protected List<Future<?>> tasks;
+
+	/** raw payload incomming */
+	protected ByteBuffer inData;
+	/** encrypted data outgoing */
+	protected ByteBuffer outCrypt;
+	/** encrypted data incoming */
+	protected ByteBuffer inCrypt;
+
+	/** the underlying channel */
+	protected SocketChannel socketChannel;
+	/** used to set interestOP SelectionKey.OP_WRITE for the underlying channel */
+	protected SelectionKey selectionKey;
+
+	protected SSLEngine sslEngine;
+	protected SSLEngineResult readEngineResult;
+	protected SSLEngineResult writeEngineResult;
+
+	/**
+	 * Should be used to count the buffer allocations.
+	 * But because of #190 where HandshakeStatus.FINISHED is not properly returned by nio wrap/unwrap this variable is used to check whether {@link #createBuffers(SSLSession)} needs to be called.
+	 **/
+	protected int bufferallocations = 0;
+
+	public SSLSocketChannel2( SocketChannel channel , SSLEngine sslEngine , ExecutorService exec , SelectionKey key ) throws IOException {
+		if( channel == null || sslEngine == null || exec == null )
+			throw new IllegalArgumentException( "parameter must not be null" );
+
+		this.socketChannel = channel;
+		this.sslEngine = sslEngine;
+		this.exec = exec;
+
+		readEngineResult = writeEngineResult = new SSLEngineResult( Status.BUFFER_UNDERFLOW, sslEngine.getHandshakeStatus(), 0, 0 ); // init to prevent NPEs
+
+		tasks = new ArrayList<Future<?>>( 3 );
+		if( key != null ) {
+			key.interestOps( key.interestOps() | SelectionKey.OP_WRITE );
+			this.selectionKey = key;
+		}
+		createBuffers( sslEngine.getSession() );
+		// kick off handshake
+		socketChannel.write( wrap( emptybuffer ) );// initializes res
+		processHandshake();
+	}
+
+	private void consumeFutureUninterruptible( Future<?> f ) {
+		try {
+			boolean interrupted = false;
+			while ( true ) {
+				try {
+					f.get();
+					break;
+				} catch ( InterruptedException e ) {
+					interrupted = true;
+				}
+			}
+			if( interrupted )
+				Thread.currentThread().interrupt();
+		} catch ( ExecutionException e ) {
+			throw new RuntimeException( e );
+		}
+	}
+
+	/**
+	 * This method will do whatever necessary to process the sslengine handshake.
+	 * Thats why it's called both from the {@link #read(ByteBuffer)} and {@link #write(ByteBuffer)}
+	 **/
+	private synchronized void processHandshake() throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 111-123, lines 125-140 into separate methods.